netplugin/nameserver/nameserver.go

```diff
@@ -0,0 +1,705 @@
+/***
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nameserver
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/Sirupsen/logrus"
+	"github.com/contiv/netplugin/core"
+	"github.com/contiv/netplugin/netmaster/mastercfg"
+	"github.com/miekg/dns"
+	cmap "github.com/streamrail/concurrent-map"
+	"hash/fnv"
+	"net"
+	"strings"
+	"sync"
+)
+
+// decorator log
+var dnsLog *logrus.Entry
+
+type tenantBucket struct {
+	sync.RWMutex
+	tenantTables map[string]*dnsTables
+}
+
+const commonK8sTenant = "common"
+
+// NetpluginNameServer config
+type NetpluginNameServer struct {
```

Comment:
not really, we can hook any nameserver in netplugin, and this file implements netplugin nameserver,
hence the name